modules/text2vec-openai/vectorizer/objects.go

```diff
@@ -99,3 +118,113 @@ func (v *Vectorizer) getVectorizationConfig(cfg moduletools.ClassConfig) ent.Vec
 		Dimensions:   settings.Dimensions(),
 	}
 }
+
+// batchWorker is a go routine that handles the communication with the vectorizer
+//
+// On the high level it has the following steps:
+//  1. It receives a batch job
+//  2. It splits the job into smaller vectorizer-batches if the token limit is reached. Note that objects from different
+//     batches are not mixed with each other.
+//  3. It sends the smaller batches to the vectorizer
+func (v *Vectorizer) batchWorker() {
+	tokenLimit := 1000 // don't know limits before the first request
+	remainingRequests := 10
+	requestLimitReset := 0
+	texts := make([]string, 0, 100)
+	for job := range v.jobQueueCh {
+		conf := v.getVectorizationConfig(job.cfg)
+		objCounter := 0
+		vecBatchOffset := 0 // index of first object in current vectorizer-batch
+
+		tokensInCurrentBatch := 0
+		texts = texts[:0]
+		for {
+			// add input to vectorizer-batches until current token limit is reached
+			text := job.texts[objCounter]
+			tokensInCurrentBatch += job.tokens[objCounter]
+			if float32(tokensInCurrentBatch) < 0.95*float32(tokenLimit) {
+				texts = append(texts, text)
+				if objCounter < len(job.objects)-1 {
+					objCounter++
+					continue
+				}
+			}
+
+			if _, ok := job.ctx.Deadline(); !ok {
+				for j := 0; j < len(texts); j++ {
+					job.errs[vecBatchOffset+j] = fmt.Errorf("context deadline exceeded")
+				}
+				break
+			}
+			res, rateLimit, err := v.client.Vectorize(job.ctx, texts, conf)
+			if err != nil {
+				for j := 0; j < len(texts); j++ {
+					job.errs[vecBatchOffset+j] = err
+				}
+			} else {
+				for j := 0; j < len(texts); j++ {
+					if res.Errors[j] != nil {
+						job.errs[vecBatchOffset+j] = res.Errors[j]
+					} else {
+						job.objects[vecBatchOffset+j].Vector = res.Vector[j]
+					}
+				}
+			}
+
+			if rateLimit != nil {
+				tokenLimit = rateLimit.RemainingTokens
+				remainingRequests = rateLimit.RemainingRequests
+				requestLimitReset = rateLimit.ResetRequests
+			}
+
+			// not all request limits are included in "RemainingRequests" and "ResetRequests". For example, in the free
+			// tier only the RPD limits are shown but not RPM
+			if remainingRequests == 0 {
+				// if we need to wait more than 60s for a reset we need to stop the batch to not produce timeouts
+				if requestLimitReset >= 61 {
+					for j := vecBatchOffset; j < len(job.objects); j++ {
+						job.errs[j] = err
+					}
+					break
+				}
+				time.Sleep(time.Duration(requestLimitReset) * time.Second)
+			}
+
+			objCounter++
+			if objCounter == len(job.objects) {
+				break
+			}
+
+			// reset for next vectorizer-batch
+			vecBatchOffset = objCounter
+			tokensInCurrentBatch = 0
+			texts = texts[:0]
+		}
+		job.wg.Done()
+
+	}
+}
+
+func (v *Vectorizer) ObjectBatch(ctx context.Context, objects []*models.Object, cfg moduletools.ClassConfig,
+) map[int]error {
+	wg := sync.WaitGroup{}
+	wg.Add(1)
```

Comment:
Since WaitGroup is waiting only for single result, it could be replaced with channel.

Reply:
what would be the upside of a channel? I thought about both and went with a waitgroup for reasons I cannot remember